### PR TITLE
Cache push notification profile image and add logs

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
This PR adds a cache for the profile images for the push notifications on iOS and fixes the delete cache in the advance settings

```release-note
NONE
```

### CODE REVIEW

File: app/utils/file/index.ts [hunk 1]
```diff
@@ -525,8 +534,18 @@ export const getAllFilesInCachesDirectory = async (serverUrl: string) => {
     try {
         const files: FileSystem.ReadDirItem[][] = [];
 
-        const directoryFiles = await FileSystem.readDir(`${FileSystem.CachesDirectoryPath}/${urlSafeBase64Encode(serverUrl)}`);
-        files.push(directoryFiles);
+        const promises = [FileSystem.readDir(`${FileSystem.CachesDirectoryPath}/${urlSafeBase64Encode(serverUrl)}`)];
+        if (Platform.OS === 'ios') {
+            const cacheDir = `${getIOSAppGroupDetails().appGroupSharedDirectory}/Library/Caches/${urlSafeBase64Encode(serverUrl)}`;
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/utils/file/index.ts:539]: Ok, the init() method of the FileCache class answers it.  But, when we download an image when viewing in the Gallery, is it also saving to the AppGroup directory for iOS?

File: ios/Gekidou/Sources/Gekidou/FileCache.swift [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  }
+  
+  private func createDirectoryIfNeeded(directory: URL?) {
+    var isDirectory = ObjCBool(false)
+       
+    if let cachePath = directory?.path {
+      let exists = FileManager.default.fileExists(atPath: cachePath, isDirectory: &isDirectory)
+      
+      if !exists && !isDirectory.boolValue {
+        try? FileManager.default.createDirectory(atPath: cachePath, withIntermediateDirectories: true, attributes: nil)
```
Comment (avinashlng1080) [inline ios/Gekidou/Sources/Gekidou/FileCache.swift:35]: it fails silently if it cannot create the cache directory

**Review by avinashlng1080 — APPROVED** [review]
Eye opening PR.  LGTM.  I just left a few questions for my own understanding.

File: app/utils/file/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (enahum) [reply app/utils/file/index.ts:539]: images from the Gallery are saved to the Photos app. Images downloaded by FastImage are stored in the app container cache.

File: ios/NotificationService/NotificationService.swift [hunk 3]
```diff
@@ -24,13 +25,24 @@ class NotificationService: UNNotificationServiceExtension {
...
       fetchReceipt(ackNotification)
     } else {
+      os_log(OSLogType.default, "Mattermost Notifications: bestAttemptContent seems to be empty, will call sendMessageIntent")
       sendMessageIntent(notification: request.content)
     }
   }
 
   func processResponse(serverUrl: String, data: Data, bestAttemptContent: UNMutableNotificationContent) {
     bestAttemptContent.userInfo["server_url"] = serverUrl
+    os_log(
```
[the remark this answers is not in the document]
Comment (enahum) [reply ios/NotificationService/NotificationService.swift:35]: As of now this are accessible through the console.app in a macOS computer.. at some point we would need to do some work to add the native logs to the logs we are storing in files.

File: ios/Gekidou/Sources/Gekidou/Networking/PushNotification.swift [hunk 4]
```diff
@@ -66,6 +67,14 @@ extension String {
         guard self.hasPrefix(prefix) else { return self }
         return String(self.dropFirst(prefix.count))
     }
+    
+    func toUrlSafeBase64Encode() -> String {
+        return Data(
+            self.replacingOccurrences(of: "/\\+/g", with: "-", options: .regularExpression)
+                .replacingOccurrences(of: "/\\//g", with: "_", options: .regularExpression)
```
[the remark this answers is not in the document]
Comment (enahum) [reply ios/Gekidou/Sources/Gekidou/Networking/PushNotification.swift:74]: is an escape character and this works perfectly fine providing the same value as the function in JS. if the string is malformed the Swift compiler yells at you.

[the remark this answers is not in the document]
Comment (enahum) [reply ios/Gekidou/Sources/Gekidou/Networking/PushNotification.swift:74]: Basically the regex is  `/\+/g` and we need to escape the `\` character so it ends up being `/\\+/g` which means match one or more `\`

**Pull request merged**